README.md
mongo-express
===

[![npm version](https://badge.fury.io/js/mongo-express.svg)](https://www.npmjs.com/package/mongo-express) [![npm](https://img.shields.io/npm/dm/mongo-express.svg)](https://www.npmjs.com/package/mongo-express) [![GitHub stars](https://img.shields.io/github/stars/mongo-express/mongo-express.svg)](https://github.com/mongo-express/mongo-express/stargazers) [![Known Vulnerabilities](https://snyk.io/test/npm/name/badge.svg)](https://snyk.io/test/npm/mongo-express)
[![Build Status](https://travis-ci.org/mongo-express/mongo-express.svg?branch=master)](https://travis-ci.org/mongo-express/mongo-express)

Web-based MongoDB admin interface written with Node.js, Express.


Features
--------

* Connect to multiple databases
* View/add/delete databases
* View/add/rename/delete collections
* View/add/update/delete documents
* Preview audio/video/image assets inline in collection view
* Nested and/or large objects are collapsible for easy overview
* Async on-demand loading of big document properties (>100KB default) to keep collection view fast
* GridFS support - add/get/delete incredibly large files
* Use BSON data types in documents
* Connect and authenticate to individual databases
* Authenticate as admin to view all databases
* Database blacklist/whitelist
* Custom CA and CA validation disabling
* Supports replica sets


Screenshots
-----------

Home Page | Database View | Collection View | Editing A Document
--- | --- | --- | ---
<img src="http://i.imgur.com/XiYhblA.png" title="Home Page showing databases"> | <img src="http://i.imgur.com/XWcIgY1.png" title="Viewing collections & buckets in a database" /> | <img src="https://imgur.com/UmGSr3x.png" title="Viewing documents in a collection" /> | <img src="https://imgur.com/lL38abn.png" title="Editing a document" />

These screenshots are from version 0.30.40
View album for more screenshots: (server status, database views etc..)
[https://imgur.com/a/9vHsF](https://imgur.com/a/9vHsF)


Development
-----------------

Copy config.default.js to config.js and edit the default property to fit your local environment

**Run the development build using:**

    npm run start-dev

Usage (npm / CLI)
-----------------

*mongo-express* requires Node.js v4 or higher.

**To install:**

    npm install -g mongo-express

Or if you want to install a non-global copy:

    npm install mongo-express

By default `config.default.js` is used where the basic access authentication is `admin`:`pass`. This is obviously not safe, and there are warnings in the console.

**To configure:**

Copy `YOUR_PATH/node_modules/mongo-express/config.default.js` into a new file called `YOUR_PATH/node_modules/mongo-express/config.js`.

*Note:* YOUR_PATH will depend on your current OS user and system configuration. You can see it in the output text shown after executing npm install.

Fill in your MongoDB connection details and any other options you want to change in `config.js`.

**To run:**


    cd YOUR_PATH/node_modules/mongo-express/ && node app.js

If you installed it globally, you can immediately start mongo-express like this:

    mongo-express -u user -p password -d database

You can access a remote database by providing MongoDB Host and Port:

    mongo-express -u user -p password -d database -H mongoDBHost -P mongoDBPort

Or if you want to use it as an administrator:

    mongo-express -a -u superuser -p password

For help on configuration options:

    mongo-express -h

Usage (Express 4 middleware)
----------------------------

**To mount as Express 4 middleware (see `node_modules/mongo-express/app.js`):**

    var mongo_express = require('mongo-express/lib/middleware')
    var mongo_express_config = require('./mongo_express_config')

    app.use('/mongo_express', mongo_express(mongo_express_config))

Usage (Docker)
--------------

Make sure you have a running [MongoDB container](https://hub.docker.com/_/mongo/) and specify it's name in the `--link` argument.

**Use the docker hub image:**

```console
$ docker run -it --rm -p 8081:8081 --link YOUR_MONGODB_CONTAINER:mongo mongo-express
```

**Build from source:**

Build an image from the project directory, then run the image.

```console
$ docker build -t mongo-express .
$ docker run -it --rm -p 8081:8081 --link YOUR_MONGODB_CONTAINER:mongo mongo-express
```

You can use the following [environment variables](https://docs.docker.com/reference/run/#env-environment-variables) to modify the container's configuration:

    Name                              | Default         | Description
    ----------------------------------|-----------------|------------
    `ME_CONFIG_MONGODB_SERVER`        |`mongo` or `localhost`| MongoDB host name or IP address. The default is `localhost` in the config file and `mongo` in the docker image. If it is a replica set, use a comma delimited list of the host names.
    `ME_CONFIG_MONGODB_PORT`          | `27017`         | MongoDB port.
    `ME_CONFIG_MONGODB_URL`           | `mongodb://admin:pass@localhost:27017/db?ssl=false`
    `ME_CONFIG_MONGODB_ENABLE_ADMIN`  | `false`         | Enable administrator access. Send strings: `"true"` or `"false"`.
    `ME_CONFIG_MONGODB_ADMINUSERNAME` | ` `             | Administrator username.
    `ME_CONFIG_MONGODB_ADMINPASSWORD` | ` `             | Administrator password.
    `ME_CONFIG_MONGODB_AUTH_DATABASE` | `db`            | Database name (only needed if `ENABLE_ADMIN` is `"false"`).
    `ME_CONFIG_MONGODB_AUTH_USERNAME` | `admin`         | Database username (only needed if `ENABLE_ADMIN` is `"false"`).
    `ME_CONFIG_MONGODB_AUTH_PASSWORD` | `pass`          | Database password (only needed if `ENABLE_ADMIN` is `"false"`).
    `ME_CONFIG_SITE_BASEURL`          | `/`             | Set the express baseUrl to ease mounting at a subdirectory. Remember to include a leading and trailing slash.
    `ME_CONFIG_SITE_COOKIESECRET`     | `cookiesecret`  | String used by [cookie-parser middleware](https://www.npmjs.com/package/cookie-parser) to sign cookies.
    `ME_CONFIG_SITE_SESSIONSECRET`    | `sessionsecret` | String used to sign the session ID cookie by [express-session middleware](https://www.npmjs.com/package/express-session).
    `ME_CONFIG_BASICAUTH_USERNAME`    | `admin`         | mongo-express web login name. Sending an empty string will disable basic authentication.
    `ME_CONFIG_BASICAUTH_PASSWORD`    | `pass`          | mongo-express web login password.
    `ME_CONFIG_REQUEST_SIZE`          | `100kb`         | Used to configure maximum mongo update payload size. CRUD operations above this size will fail due to restrictions in [body-parser](https://www.npmjs.com/package/body-parser).
    `ME_CONFIG_OPTIONS_EDITORTHEME`   | `rubyblue`      | Web editor color theme, [more here](http://codemirror.net/demo/theme.html).
    `ME_CONFIG_SITE_SSL_ENABLED`      | `false`         | Enable SSL.
    `ME_CONFIG_MONGODB_SSLVALIDATE`   | `true`          | Validate mongod server certificate against CA
    `ME_CONFIG_SITE_SSL_CRT_PATH`     | ` `             | SSL certificate file.
    `ME_CONFIG_SITE_SSL_KEY_PATH`     | ` `             | SSL key file.
    `ME_CONFIG_SITE_GRIDFS_ENABLED`   | `false`         | Enable gridFS to manage uploaded files.

**Example:**

    docker run -it --rm \
        --name mongo-express \
        --link web_db_1:mongo \
        -p 8081:8081 \
        -e ME_CONFIG_OPTIONS_EDITORTHEME="ambiance" \
        -e ME_CONFIG_BASICAUTH_USERNAME="" \
        mongo-express

This example links to a container name typical of `docker-compose`, changes the editor's color theme, and disables basic authentication.

**To use:**

The default port exposed from the container is 8081, so visit `http://localhost:8081` or whatever URL/port you entered into your config (if running standalone) or whatever `config.site.baseUrl` (if mounting as a middleware).

Usage (Bluemix)
---------------

**Deploy to Bluemix**

Doing manually:

* Git clone this repository
* Create a new or use already created [MongoDB experimental service](https://www.ng.bluemix.net/docs/#services/MongoDB/index.html#MongoDB)
* Change the file `manifest.yml` to fit your Bluemix app and service environment


Doing automatically:

* Click the button below to fork into IBM DevOps Services and deploy your own copy of this application on Bluemix

[![Deploy to Bluemix](https://bluemix.net/deploy/button.png)](https://bluemix.net/deploy?repository=https://github.com/mongo-express/mongo-express.git)


Then, take the following action to customize to your environment:

* Create your `config.js` file based on `config.default.js`
  * Check if it is necessary to change the `dbLabel` according to the MongoDB service created
  * Change the `basicAuth` properties, not to keep the default values


Search
------
* *Simple* search takes the user provided fields (`key` & `value`) and prepares a MongoDB find() object, with projection set to `{}` so returns all columns.
* *Advanced* search passes the `find` and `projection` fields/objects straight into MongoDB `db.collection.find(query, projection)`. The `find` object is where your query happens, while the `projection` object determines which columns are returned.

See [MongoDB db.collection.find()](https://docs.mongodb.org/manual/reference/method/db.collection.find/) documentation for examples and exact usage.

Planned features
----------------

Pull Requests are always welcome! <3


Limitations
-----------

* Documents must have `document._id` property to be edited
* Binary BSON data type not tested

Not tested
----------

* Binary/BinData

JSON documents are parsed through a javascript virtual machine, so **the web
interface can be used for executing malicious javascript on a server**.

**mongo-express should only be used privately for development purposes**.



BSON Data Types
---------------

The following BSON data types are supported in the mongo-express document editor/viewer.

**Native Javascript Types**

Strings, numbers, lists, booleans, null, etc.

All numbers in Javascript are 64-bit floating points.

**ObjectID/ObjectId**

    ObjectID()

Creates a new Object ID type.

    ObjectID(id)

Use Object ID with the given 24-digit hexadecimal string.

**ISODate**

    ISODate()

Creates a new ISODate object with current time.

`new Date()` can also be used (note the `new` keyword there).

    ISODate(timestamp)

Uses ISODate object with the given timestamp.

**DBRef/Dbref**

    DBRef(collection, objectID)

    DBRef(collection, objectID, database)

Object ID is the ID string, not the ObjectID type.

The database value is optional.

**Timestamp**

    Timestamp()

Creates a new Timestamp object with a value of 0.

    Timestamp(time, ordinal)

Example: `Timestamp(ISODate(), 0)`.

See [http://www.mongodb.org/display/DOCS/Timestamp+data+type](http://www.mongodb.org/display/DOCS/Timestamp+data+type) for more info about the Timestamp data type.

**Code**

    Code(code)

Code can be a native Javascript function, or it can be a string.

Specifying a scope/context is not supported.

**MinKey**

    MinKey()

**MaxKey**

    MaxKey()

**Symbol**

    Symbol(string)

Example Document
----------------

Here is an example of a document which can be read/edited in mongo-express (media truncated for legibility):

    {
      "_id": ObjectID(), // or ObjectId()
      "dates": {
        "date": ISODate("2012-05-14T16:20:09.314Z"),
        "new_date": ISODate(),
        "alternative": new Date()
      },
      "photo": "data:image/jpeg;base64,/9j/4...",
      "video": "data:video/webm;base64,GkXfo...",
      "audio": "data:audio/ogg;base64,T2dnUw...",
      "bool": true,
      "string": "hello world!",
      "list of numbers": [
        123,
        111e+87,
        4.4,
        -12345.765
      ],
      "reference": DBRef("collection", "4fb1299686a989240b000001"),
      "ts": Timestamp(ISODate(), 1),
      "minkey": MinKey(),
      "maxkey": MaxKey(),
      "func": Code(function() { alert('Hello World!') }),
      "symbol": Symbol("test")
    }

---

License
-------
MIT License

Copyright (c) 2012 Chun-hao Hu
Copyright (c) 2016-present Multiple Contributors


Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


HISTORY.md
0.42.2 - Sat Aug 12, 2017
------

- Fixed default port in lib/db.js from documentation


0.42.1 - Mon Aug 7, 2017
------

- Fixed typo in tag of collection view (#357)


0.42.0 - Thu Jul 20, 2017
------

- Add connectionString support to connection (#350)


0.41.0 - Tue Jul 4, 2017
------

- Use stream for exporting collections (#345)
- Solution for CSRF wiht base url (#347)


0.40.0 - Mon May 1, 2017
------

- Fixed more flexible db names (#339)
- Fixed escaping in nested objects
- Fixed path resolve for windows for the builds
- Fixed revise the list database function to support user having access to admin database (#338)


0.39.2 - Sun Apr 16, 2017
------

## Fixes
* Fixed bug with collapsibleJSON property not able to render properly


0.39.1 - Sun Mar 26, 2017
------

## New
* Better windows support with usage of cross-env


0.39.0 - Sun Mar 26, 2017
------

## New 
* Add Create Index functionality (#326)

## Chores
* Added a bit of readme info, and updated license


0.38.0 - Sat Feb 25, 2017
------

## New
* Support for mounting on a different baseHref (#320)

## Chores
* Improve configuration step (#321)


0.37.2 - Tue Feb 21, 2017
------

## Fixes
* Small publishing fixes


0.37.1 - Tue Feb 21, 2017
------

## Fixes
* Build is now shipped as part of package


0.37.0 - Mon Feb 20, 2017
------

## New
* Added support for Collection with slash in the name (#309)
* Changed the way scripts work (#312)
    - Scripts are now built using babel and are in their own files.
    - Each page has it's own mini-bundle + vendor bundle
    - Scripts are built on install and output a json file that is used for url purposes
* Added more tests, and new test utils! (#314, #315)

## Fixes
* Fixed issue where parenthese we're stringified incorrectly (#306)
* Fixed loadDocument script that was rendered properly in read-only (#308)

## Varia
* Removed underscore and use only lodash. (#307)
* Removed Snyk. It was causing more annoyance than actual good (#306)
* Updated swig to swig-template, which uses a safe version (#311)


0.36.0 - Tue Feb 7, 2017
------
* Fix content-disposition non-ascii header, set config.gridFSEnabled via env vars (#304)

0.35.0 - Thu Jan 5, 2017
------
* Removed Duplicated navbar in mobile layout (#302)
* Add logger options to control logging (#300)
* Pass options in middleware-only usage (#299)

0.34.0 - Fri Dec 23, 2016
------

## Fixes
* Fixed index are no longer deletable in read-only mode (#298)

0.33.0 - Tue Dec 6, 2016
------

## New 
* Delete all Documents matching a query.

## Fixes

* Fixed issue where binary types we're converted to string when updating a document

0.32.0 - Sat Nov 12, 2016
------

## New
* ObjectID can now be used without the ObjectID wrapper.
* Added export to CSV
* Added index management. You can now delete index or reIndex a collection

## Fixes
* Fixed issue where ObjectID we're incorrectly stringified


0.30.48
-------
 * Add support for super large Objects

0.21.0
------

* Added database statistics (karthik25)
* Added basic auth (netpi)
* Added complex querying (kathik25)

0.20.0
------

* Added JSON find type
* Added collection export
* Added confirmation dialog on delete
* Added uptime info in days
* Fixed long collection name issue

0.18.0
------

* Updated express package version
* Updated swig package version
* Added simple key-value document searching

0.17.5
------

* Specified version 3.0.0alpha1 of express in package.json. Latest version of express is causing errors.

0.17.4
------

* Removed hacky BSON->string conversion function
* Replaced it with a modified JSON.stringify function

0.17.3
------

* Removed requirement for doc IDs to be ObjectID type

0.17.2
------

* Added build status images in README

0.17.1
------

* Added tests for BSON conversion functions

0.17.0
------

* Added support for all BSON data types except Binary
* Fixed BSON converter so not only top-level values get converted
* Updated README with more BSON data type examples

0.16.1
------

* Fixed bug: when trying to delete document, collection gets deleted

0.16.0
------

* Added support for some BSON data types when viewing docs
* Updated README with list of supported data types

0.15.0
------

* Added support for BSON data types when adding/editing docs

0.14.1
------

* Forgot to update HISTORY file

0.14.0
------

* Added success/error feedback messages
* Added cookie/session middleware
* Added cookieSecret and sessionSecret options to config
* Moved config.js to config.default.js
* Updated configuration instructions in README

0.13.0
------

* Added version history
* Added MIT license
* Added pager links above document list
* Added pagination links below document list
* Added config option to set number of documents displayed per page

0.12.0
------

* Added async package to dependencies
* Added system.users to list of un-editable collections
* Changed config file to allow lists of databases and associated auth details
* Added support for regular MongoDB users
* Removed requirement for admin access to the MongoDB server


config.default.js
'use strict';

var mongo;

// Accesing Bluemix variable to get MongoDB info
if (process.env.VCAP_SERVICES) {
  var dbLabel = 'mongodb-2.4';
  var env = JSON.parse(process.env.VCAP_SERVICES);
  if (env[dbLabel]) {
    mongo = env[dbLabel][0].credentials;
  }
} else {
  mongo = {
    // setting the connection string will only give access to that database
    // to see more databases you need to set mongodb.admin to true or add databases to the mongodb.auth list
    connectionString: process.env.ME_CONFIG_MONGODB_SERVER ? '' : process.env.ME_CONFIG_MONGODB_URL,
  };
}

var meConfigMongodbServer = process.env.ME_CONFIG_MONGODB_SERVER ? process.env.ME_CONFIG_MONGODB_SERVER.split(',') : false;

module.exports = {
  mongodb: {
    // if a connection string options such as server/port/etc are ignored
    connectionString: mongo.connectionString || '',

    //server: mongodb hostname or IP address
    //for replica set, use array of string instead
    server: (meConfigMongodbServer.length > 1 ? meConfigMongodbServer : meConfigMongodbServer[0]) || mongo.host,
    port:   process.env.ME_CONFIG_MONGODB_PORT || mongo.port,

    //ssl: connect to the server using secure SSL
    ssl: process.env.ME_CONFIG_MONGODB_SSL || mongo.ssl,

    //sslValidate: validate mongod server certificate against CA
    sslValidate: process.env.ME_CONFIG_MONGODB_SSLVALIDATE || true,
    

    //sslCA: array of valid CA certificates
    sslCA:  [],

    //autoReconnect: automatically reconnect if connection is lost
    autoReconnect: true,

    //poolSize: size of connection pool (number of connections to use)
    poolSize: 4,

    //set admin to true if you want to turn on admin features
    //if admin is true, the auth list below will be ignored
    //if admin is true, you will need to enter an admin username/password below (if it is needed)
    admin: process.env.ME_CONFIG_MONGODB_ENABLE_ADMIN ? process.env.ME_CONFIG_MONGODB_ENABLE_ADMIN.toLowerCase() === 'true' : false,

    // >>>>  If you are using regular accounts, fill out auth details in the section below
    // >>>>  If you have admin auth, leave this section empty and skip to the next section
    auth: [
      /*
       * Add the name, username, and password of the databases you want to connect to
       * Add as many databases as you want!
       */
      {
        database: process.env.ME_CONFIG_MONGODB_AUTH_DATABASE || mongo.db,
        username: process.env.ME_CONFIG_MONGODB_AUTH_USERNAME || mongo.username,
        password: process.env.ME_CONFIG_MONGODB_AUTH_PASSWORD || mongo.password,
      },
    ],

    //  >>>>  If you are using an admin mongodb account, or no admin account exists, fill out section below
    //  >>>>  Using an admin account allows you to view and edit all databases, and view stats

    //leave username and password empty if no admin account exists
    adminUsername: process.env.ME_CONFIG_MONGODB_ADMINUSERNAME || '',
    adminPassword: process.env.ME_CONFIG_MONGODB_ADMINPASSWORD || '',

    //whitelist: hide all databases except the ones in this list  (empty list for no whitelist)
    whitelist: [],

    //blacklist: hide databases listed in the blacklist (empty list for no blacklist)
    blacklist: [],
  },

  site: {
    // baseUrl: the URL that mongo express will be located at - Remember to add the forward slash at the start and end!
    baseUrl: process.env.ME_CONFIG_SITE_BASEURL || '/',
    cookieKeyName: 'mongo-express',
    cookieSecret:     process.env.ME_CONFIG_SITE_COOKIESECRET   || 'cookiesecret',
    host:             process.env.VCAP_APP_HOST                 || 'localhost',
    port:             process.env.VCAP_APP_PORT                 || 8081,
    requestSizeLimit: process.env.ME_CONFIG_REQUEST_SIZE        || '50mb',
    sessionSecret:    process.env.ME_CONFIG_SITE_SESSIONSECRET  || 'sessionsecret',
    sslCert:          process.env.ME_CONFIG_SITE_SSL_CRT_PATH   || '',
    sslEnabled:       process.env.ME_CONFIG_SITE_SSL_ENABLED    || false,
    sslKey:           process.env.ME_CONFIG_SITE_SSL_KEY_PATH   || '',
  },

  //set useBasicAuth to true if you want to authenticate mongo-express loggins
  //if admin is false, the basicAuthInfo list below will be ignored
  //this will be true unless ME_CONFIG_BASICAUTH_USERNAME is set and is the empty string
  useBasicAuth: process.env.ME_CONFIG_BASICAUTH_USERNAME !== '',

  basicAuth: {
    username: process.env.ME_CONFIG_BASICAUTH_USERNAME || 'admin',
    password: process.env.ME_CONFIG_BASICAUTH_PASSWORD || 'pass',
  },

  options: {
    // Display startup text on console
    console: true,

    //documentsPerPage: how many documents you want to see at once in collection view
    documentsPerPage: 10,

    //editorTheme: Name of the theme you want to use for displaying documents
    //See http://codemirror.net/demo/theme.html for all examples
    editorTheme: process.env.ME_CONFIG_OPTIONS_EDITORTHEME || 'rubyblue',

    // Maximum size of a single property & single row
    // Reduces the risk of sending a huge amount of data when viewing collections
    maxPropSize: (100 * 1000),  // default 100KB
    maxRowSize: (1000 * 1000),  // default 1MB

    //The options below aren't being used yet

    //cmdType: the type of command line you want mongo express to run
    //values: eval, subprocess
    //  eval - uses db.eval. commands block, so only use this if you have to
    //  subprocess - spawns a mongo command line as a subprocess and pipes output to mongo express
    cmdType: 'eval',

    //subprocessTimeout: number of seconds of non-interaction before a subprocess is shut down
    subprocessTimeout: 300,

    //readOnly: if readOnly is true, components of writing are not visible.
    readOnly: false,

    //collapsibleJSON: if set to true, jsons will be displayed collapsible
    collapsibleJSON: true,

    //collapsibleJSONDefaultUnfold: if collapsibleJSON is set to `true`, this defines default level
    //  to which JSONs are displayed unfolded; use number or "all" to unfold all levels
    collapsibleJSONDefaultUnfold: 1,

    //gridFSEnabled: if gridFSEnabled is set to 'true', you will be able to manage uploaded files ( ak. grids, gridFS )
    gridFSEnabled: process.env.ME_CONFIG_SITE_GRIDFS_ENABLED || false,

    // logger: this object will be used to initialize router logger (morgan)
    logger: {},
  },

  // Specify the default keyname that should be picked from a document to display in collections list.
  // Keynames can be specified for every database and collection.
  // If no keyname is specified, it defaults to '_id', which is a mandatory field.
  // For Example :
  // defaultKeyNames{
  //   "world_db":{  //Database Name
  //     "continent":"cont_name", // collection:field
  //     "country":"country_name",
  //     "city":"name"
  //   }
  // }
  defaultKeyNames: {

  },
};
